diff --git a/net/ipv4/netfilter/ip_tables.c b/net/ipv4/netfilter/ip_tables.c
index 10b91ebdf2131..05d89f182bbe6 100644
--- a/net/ipv4/netfilter/ip_tables.c
+++ b/net/ipv4/netfilter/ip_tables.c
@@ -220,6 +220,16 @@ struct ipt_entry *ipt_next_entry(const struct ipt_entry *entry)
 	return (void *)entry + entry->next_offset;
 }
 
+static inline
+struct ipt_entry *ipt_n_entry(struct ipt_entry *entry, u_int8_t count)
+{
+	do {
+		entry = (void *)entry + entry->next_offset;
+	} while(count--);
+	
+	return (void*)entry;
+}
+
 /* Returns one of the generic firewall policies, like NF_ACCEPT. */
 unsigned int
 ipt_do_table(struct sk_buff *skb,
@@ -233,7 +243,7 @@ ipt_do_table(struct sk_buff *skb,
 	unsigned int verdict = NF_DROP;
 	const char *indev, *outdev;
 	const void *table_base;
-	struct ipt_entry *e, **jumpstack;
+	struct ipt_entry *e, *tmpe, **jumpstack;
 	unsigned int stackidx, cpu;
 	const struct xt_table_info *private;
 	struct xt_action_param acpar;
@@ -348,6 +358,15 @@ ipt_do_table(struct sk_buff *skb,
 			/* Target might have changed stuff. */
 			ip = ip_hdr(skb);
 			e = ipt_next_entry(e);
+		} else if(verdict & 0xFF) {
+			verdict = verdict >> 8;
+			tmpe = ipt_n_entry(e, verdict);
+			if(tmpe == e) {
+				verdict = NF_ACCEPT;
+				break;
+			}
+			e = tmpe;
+			verdict = XT_CONTINUE;
 		} else {
 			/* Verdict */
 			break;
